pkg/standalone-utils/contracts/BalancerSorQueries.sol

```diff
@@ -0,0 +1,293 @@
+        }
+
+        if (config.loadLinearWrappedTokenRates) {
+            address[] memory linearPools = new address[](config.linearPoolIdxs.length);
+
+            for (i = 0; i < config.linearPoolIdxs.length; i++) {
+                linearPools[i] = pools[config.linearPoolIdxs[i]];
+            }
+
+            linearWrappedTokenRates = getWrappedTokenRateForLinearPools(linearPools);
+        }
+
+        if (config.loadNormalizedWeights) {
+            address[] memory weightedPools = new address[](config.weightedPoolIdxs.length);
+
+            for (i = 0; i < config.weightedPoolIdxs.length; i++) {
+                weightedPools[i] = pools[config.weightedPoolIdxs[i]];
+            }
+
+            weights = getNormalizedWeightsForPools(weightedPools);
+        }
+
+        if (config.loadScalingFactors) {
+            address[] memory scalingFactorPools = new address[](config.scalingFactorPoolIdxs.length);
+
+            for (i = 0; i < config.scalingFactorPoolIdxs.length; i++) {
+                scalingFactorPools[i] = pools[config.scalingFactorPoolIdxs[i]];
+            }
+
+            scalingFactors = getScalingFactorsForPools(scalingFactorPools);
+        }
+
+        if (config.loadAmps) {
+            address[] memory ampPools = new address[](config.ampPoolIdxs.length);
+
+            for (i = 0; i < config.ampPoolIdxs.length; i++) {
+                ampPools[i] = pools[config.ampPoolIdxs[i]];
+            }
+
+            amps = getAmpForPools(ampPools);
+        }
+    }
+
+    function getPoolTokenBalancesWithUpdatesAfterBlock(bytes32[] memory poolIds, uint256 blockNumber)
+        public
+        view
+        returns (uint256[][] memory)
+    {
+        uint256[] memory balances;
+        uint256 lastChangeBlock;
+        uint256[][] memory allBalances = new uint256[][](poolIds.length);
+
+        for (uint256 i = 0; i < poolIds.length; i++) {
+            (, balances, lastChangeBlock) = vault.getPoolTokens(poolIds[i]);
+
+            if (lastChangeBlock > blockNumber) {
+                allBalances[i] = balances;
+            }
+        }
+
+        return allBalances;
+    }
+
+    function getWrappedTokenRateForLinearPools(address[] memory poolAddresses) public view returns (uint256[] memory) {
+        uint256[] memory rates = new uint256[](poolAddresses.length);
+
+        for (uint256 i = 0; i < poolAddresses.length; i++) {
+            rates[i] = _getLinearWrappedTokenRate(poolAddresses[i]);
+        }
+
+        return rates;
+    }
+
+    function getAmpForPools(address[] memory poolAddresses) public view returns (uint256[] memory) {
+        uint256[] memory amps = new uint256[](poolAddresses.length);
+
+        for (uint256 i = 0; i < poolAddresses.length; i++) {
+            (amps[i], , ) = IPoolWithAmp(poolAddresses[i]).getAmplificationParameter();
+        }
+
+        return amps;
+    }
+
+    function getSwapFeePercentageForPools(address[] memory poolAddresses, SwapFeeType[] memory swapFeeTypes)
```

Comment:
I'd consider checking that these arrays have the same length, to help detect user error

Reply:
totally fair, will add that!